hyperspy/model.py

```diff
@@ -960,60 +946,20 @@ def reset_the_signal_range(self):
         self.update_plot()
 
     def _model_function(self, param):
-
-        if self.convolved is True:
-            counter = 0
-            sum_convolved = np.zeros(len(self.convolution_axis))
-            sum = np.zeros(len(self.axis.axis))
-            for component in self:  # Cut the parameters list
-                if component.active:
-                    if component.convolved is True:
-                        np.add(sum_convolved, component.__tempcall__(param[
-                            counter:counter + component._nfree_param],
-                            self.convolution_axis), sum_convolved)
-                    else:
-                        np.add(
-                            sum,
-                            component.__tempcall__(
-                                param[
-                                    counter:counter +
-                                    component._nfree_param],
-                                self.axis.axis),
-                            sum)
-                    counter += component._nfree_param
-
-            to_return = (sum + np.convolve(self.low_loss(self.axes_manager),
-                                           sum_convolved, mode="valid"))[
-                self.channel_switches]
-
-        else:
-            axis = self.axis.axis[self.channel_switches]
-            counter = 0
-            first = True
-            for component in self:  # Cut the parameters list
-                if component.active:
-                    if first is True:
-                        sum = component.__tempcall__(
-                            param[
-                                counter:counter +
-                                component._nfree_param],
-                            axis)
-                        first = False
-                    else:
-                        sum += component.__tempcall__(
-                            param[
-                                counter:counter +
-                                component._nfree_param],
-                            axis)
-                    counter += component._nfree_param
-            to_return = sum
-
-        if self.spectrum.metadata.Signal.binned is True:
-            to_return *= self.spectrum.axes_manager[-1].scale
+        old_p0 = self.p0
+        self._set_p0()
+        old_values = self.p0
+        self.p0 = param
+        self._fetch_values_from_p0()
+        to_return = self.__call__(non_convolved=False, onlyactive=True)
+        self.p0 = old_values
+        self._fetch_values_from_p0()
```

Comment:
All this play was only to, in the end of the function call, have the same `self.p0` and all `values`. If it's actually not necessary (i'm not sure if we break anything), then yes, only fetching the `param` values should be enough.

I'll try to make it so and see if anything breaks